src/mysql_protocol.erl

```diff
@@ -1230,37 +1294,93 @@ get_null_terminated_binary(<<0, Rest/binary>>, Acc) ->
 get_null_terminated_binary(<<Ch, Rest/binary>>, Acc) ->
     get_null_terminated_binary(Rest, <<Acc/binary, Ch>>).
 
--spec hash_password(Password :: iodata(), Salt :: binary()) -> Hash :: binary().
-hash_password(Password, Salt) ->
+-spec hash_password(AuthMethod :: binary(), Password :: iodata(), Salt :: binary()) -> Hash :: binary().
+hash_password(AuthMethod, Password, Salt) when not is_binary(Password) ->
```

Comment:
`when is_list(Password)` -- better check what it is rather than what it isn't. (It's legacy, I know, but there is no risk of breaking anything here where iolist_to_binary is used on it.)

Reply:
Maybe even better, we could normalize those things to binaries in the entry functions like `handshake`, and just assume we work with binaries in all the internal functions?

Reply:
Sure, for username and password and such things, if it reduces complexity.

For queries and other potential large strings, we better not. The IO functions can send nested lists to sockets by just traversing them, without allocating a binary.

Reply:
Yeah, I was thinking of things like username and password, things we need to work with internally, so we don't have to do normalization in multiple places. For others like queries that we just pass along, there is no gain.

In reply to the original request, I turned it around a little in the last commit and check with `is_binary`, since this is really what we _want_ it to be to work with.

Reply:
Yes, do this if you want. It can also be done later. This PR is almost ready to merge, right? Whenever you remove the "WIP"...

Reply:
I think it is ready now, and so I remove the WIP :)